### PR TITLE
transmission, transmission_4: set mainProgram and restore nixosTests

### PR DESCRIPTION
**Pull request by kirillrdy**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/p2p/transmission/default.nix [hunk 1]
```diff
@@ -33,28 +34,21 @@
...
+stdenv.mkDerivation {
   pname = "transmission";
   inherit version;
 
   src = fetchFromGitHub {
     owner = "transmission";
     repo = "transmission";
     rev = version;
-    sha256 = "0ccg0km54f700x9p0jsnncnwvfnxfnxf7kcm7pcx1cj0vw78924z";
+    sha256 = "sha256-Sz3+5VvfOgET1aiormEnBOrF+yN79tiSQvjLAoGqTLw=";
```
Comment (eclairevoyant) [inline pkgs/applications/networking/p2p/transmission/default.nix:48]:
```suggestion
    hash = "sha256-Sz3+5VvfOgET1aiormEnBOrF+yN79tiSQvjLAoGqTLw=";
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Comment by jtojnar** [comment]
We already have it since https://github.com/NixOS/nixpkgs/pull/241146

**Comment by kirillrdy** [comment]
> We already have it since #241146. Also note there are data-loss issues: [#215316 (comment)](https://github.com/NixOS/nixpkgs/pull/215316#issuecomment-1485097961)

ah thanks, github search :'(

I'll de-scope this to just mainProgram change

File: pkgs/applications/networking/p2p/transmission/4.nix [hunk 2]
```diff
@@ -141,8 +142,14 @@ stdenv.mkDerivation rec {
...
   '';
 
+  passthru.tests = {
+    apparmor = nixosTests.transmission_4; # starts the service with apparmor enabled
+    smoke-test = nixosTests.bittorrent;
+  };
+
   meta = {
     description = "A fast, easy and free BitTorrent client";
+    mainProgram = (transmission.override { inherit enableQt enableGTK3 enableCli; }).meta.mainProgram;
```
Comment (jtojnar) [inline pkgs/applications/networking/p2p/transmission/4.nix:152]: We should not couple expressions for 4.x and 3.x to allow for independent evolution.

File: nixos/tests/all-tests.nix [hunk 3]
```diff
@@ -816,7 +816,8 @@ in {
   tor = handleTest ./tor.nix {};
   traefik = handleTestOn ["aarch64-linux" "x86_64-linux"] ./traefik.nix {};
   trafficserver = handleTest ./trafficserver.nix {};
-  transmission = handleTest ./transmission.nix {};
+  transmission = handleTest ./transmission.nix { transmission = pkgs.transmission; };
+  transmission_4 = handleTest ./transmission.nix { transmission = pkgs.transmission_4; };
```
Comment (kirillrdy) [inline nixos/tests/all-tests.nix:820]: i am happy to drop this if people feel like there isn't much value

**Review by bjornfor — APPROVED** [review]

**Comment by doronbehar** [comment]
You all might be interested in https://github.com/NixOS/nixpkgs/pull/258058 .

File: pkgs/applications/networking/p2p/transmission/4.nix [hunk 4]
```diff
@@ -141,8 +141,14 @@ stdenv.mkDerivation rec {
     EOF
   '';
 
+  passthru.tests = {
+    apparmor = nixosTests.transmission_4; # starts the service with apparmor enabled
+    smoke-test = nixosTests.bittorrent;
```
Comment (jtojnar) [inline pkgs/applications/networking/p2p/transmission/4.nix:146]: Is not this testing Transmission 3?

Comment (kirillrdy) [reply pkgs/applications/networking/p2p/transmission/4.nix:146]: this is transmission_4, I assume we want to test transmission_4 here

**Comment by kirillrdy** [comment]
> You all might be interested in #258058 .

Yes !!! while I know 4 has a regression, I think it will be less confusing and inline with other distributions, to have 4 as default.

I am happy to put this in draft and wait for rename to happen

**Comment by jtojnar** [comment]
> I think it will be less confusing and inline with other distributions, to have 4 as default.

It will also delete people’s data unless they are aware of the bug and switch to `transmission_3` before running transmission.

**Comment by kirillrdy** [comment]
> > I think it will be less confusing and inline with other distributions, to have 4 as default.
> 
> It might also delete people’s data unless they are aware of the bug and switch to `transmission_3` before running transmission.

Can't find in all other threads, have you tried 4.0.4 ? I just tried downloading a torrent using 3, wait for it to start seeding. upgrade to 4.0.4 and file is still seeding for me

**Review by doronbehar — APPROVED** [review]
Diff LGTM. CI is green.

**Pull request merged**